File: 501-find-mode-in-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.counter = {}

        def visit(root):
            if root is None:
                return
            
            if root.val not in self.counter:
                self.counter[root.val] = 1
            else:
                self.counter[root.val] += 1
            
            visit(root.left)
            visit(root.right)
        
        visit(root)
        max_count = max(self.counter.values())
        res = [
            k for k, v in self.counter.items() if v == max_count
        ]
        return res

File: 501-find-mode-in-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> result;
    int maxFreq = 0, current = 0, freq = 0;
    
    vector<int> findMode(TreeNode* root) {
        // base condition
        if (!root) return result;

        // traverse left sub tree    
        findMode(root->left);
        
        // upon visiting a node
        if (current != root->val) {
           freq = 1;
           current = root->val;
        } else if (current == root->val) {
            freq ++;   
        }
        
        if (freq > maxFreq) {
            maxFreq = freq;
            result = {current};
        } else if (freq == maxFreq) {
            result.push_back(current);
        }
        // traverse right subtree
        findMode(root->right);   
        return result;
    }
};
